e2e/nomostest/clean.go

```diff
@@ -126,6 +133,88 @@ func Clean(nt *NT) error {
 	return deleteImplicitNamespacesAndWait(nt)
 }
 
+func uninstallAllPackages(nt *NT) error {
+	// Check if reconciler-manager is reconciled and healthy
+	nt.T.Log("[CLEANUP] checking reconciler-manager status...")
+	rmObj := &unstructured.Unstructured{}
+	rmObj.SetGroupVersionKind(kinds.Deployment())
+	rmObj.SetNamespace(configmanagement.ControllerNamespace)
+	rmObj.SetName(reconcilermanager.ManagerName)
+	if err := nt.KubeClient.Get(rmObj.GetName(), rmObj.GetNamespace(), rmObj); err != nil {
+		return fmt.Errorf("checking reconciler-manager status: %v", err)
+	}
+	if err := testpredicates.StatusEquals(nt.Scheme, kstatus.CurrentStatus)(rmObj); err != nil {
+		return fmt.Errorf("reconciler-manager not healthy: %v", err)
+	}
+	// If reconciler-manager is healthy, uninstall all the unmanaged RootSyncs & RepoSyncs.
+	// If deletion-propagation is enabled, this will also recursively delete all
+	// managed objects, including other managed packages.
+	if err := uninstallUnmanagedPackagesOfType(nt, kinds.RepoSyncV1Beta1()); err != nil {
+		return fmt.Errorf("uninstalling unmanged RepoSyncs: %v", err)
+	}
+	if err := uninstallUnmanagedPackagesOfType(nt, kinds.RootSyncV1Beta1()); err != nil {
+		return fmt.Errorf("uninstalling unmanged RootSyncs: %v", err)
+	}
+	return nil
+}
+
+func uninstallUnmanagedPackagesOfType(nt *NT, gvk schema.GroupVersionKind) error {
+	// Find all packages
+	nt.T.Logf("[CLEANUP] listing %s objects...", gvk.Kind)
+	rsObjList := &unstructured.UnstructuredList{}
+	rsObjList.SetGroupVersionKind(kinds.ListGVKForItemGVK(gvk))
+	if err := nt.KubeClient.List(rsObjList, client.InNamespace(configmanagement.ControllerNamespace)); err != nil {
+		return fmt.Errorf("listing %s: %v", gvk.Kind, err)
+	}
+	// Find the unmanaged packages
+	var rsObjs []*unstructured.Unstructured
+	for _, item := range rsObjList.Items {
+		if core.GetAnnotation(&item, metadata.ResourceManagementKey) == metadata.ResourceManagementEnabled {
+			rsObjs = append(rsObjs, &item)
+		}
+	}
+	// Delete the unmanaged packages in parallel
+	nt.T.Log("[CLEANUP] deleting unmanaged %s objects...", gvk.Kind)
+	tg := taskgroup.New()
+	for _, obj := range rsObjs {
+		nn := client.ObjectKeyFromObject(obj)
+		gvk, err := kinds.Lookup(obj, nt.Scheme)
+		if err != nil {
+			return err
+		}
+		if !obj.GetDeletionTimestamp().IsZero() {
+			nt.T.Logf("[CLEANUP] already terminating %s object %s ...", gvk.Kind, nn)
+		} else {
+			nt.T.Logf("[CLEANUP] deleting %s object %s ...", gvk.Kind, nn)
+			if err := nt.KubeClient.Delete(obj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+				if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+					// skip waiting
+					continue
+				}
+				return fmt.Errorf("unable to delete %s object %s: %w",
```

Comment:
The early return here and on line 183 means we wouldn't wait for any other elements of the `taskgroup` to complete before returning. I think we want to continue and defer returning the error, so that we wait for other goroutines to complete before exiting the function.

This is what I have in mind to elegantly group the errors:
```golang
			if err := nt.KubeClient.Delete(obj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
				if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
					// skip waiting
					continue
				}
                tg.Go(func() error {
				  return fmt.Errorf("unable to delete %s object %s: %w",
					  gvk.Kind, nn, err)
                })
                continue
			}
```

Reply:
Done.